### PR TITLE
Specify the margin in price lists

### PR DESCRIPTION
**Pull request by arek1-618**
Specify the margin in price lists

chrome_2019-06-17_13-29-21

### CODE REVIEW

File: public_html/layouts/basic/modules/Vtiger/resources/RelatedList.js [hunk 1]
```diff
@@ -37,6 +37,44 @@ jQuery.Class(
 			}
 			return Vtiger_RelatedList_Js.relatedListInstance;
 		},
+		triggerMassMargin: function() {
```
Comment (TomaszQr) [inline public_html/layouts/basic/modules/Vtiger/resources/RelatedList.js]: move to PriceBooks_RelatedList_Js

File: modules/PriceBooks/actions/RelationAjax.php [hunk 2]
```diff
@@ -10,20 +10,55 @@
...
+		$queryGenerator->setField('purchase');
+		$currencyId = Vtiger_Record_Model::getInstanceById($recordId, 'PriceBooks')->get('currency_id');
+		$dbCommand = \App\Db::getInstance()->createCommand();
+		$dataReader = $queryGenerator->createQuery()->createCommand()->query();
+		while ($row = $dataReader->read()) {
+			if (!\App\Json::isEmpty($row['purchase'])) {
+				$purchase = \App\Json::decode($row['purchase']);
+				if (isset($purchase['currencies'][$currencyId]['price'])) {
+					$price = ((100.00 + $margin) / 100.00) * (float) $purchase['currencies'][$currencyId]['price'];
+					$dbCommand->update('vtiger_pricebookproductrel', ['listprice' => $price], ['productid' => $row['id']])->execute();
```
Comment (TomaszQr) [inline modules/PriceBooks/actions/RelationAjax.php]: pricebookid ?

File: modules/PriceBooks/actions/RelationAjax.php [hunk 3]
```diff
@@ -10,20 +10,55 @@
...
+	 */
+	public function specifyMargin(App\Request $request)
 	{
-		$mode = $request->getMode();
-		if (!empty($mode) && method_exists($this, "$mode")) {
-			$this->$mode($request);
+		$margin = $request->getByType('margin', 'Double');
+		$recordId = $request->getInteger('record');
+		$queryGenerator = static::getQuery($request);
+		$queryGenerator->setField('purchase');
```
Comment (TomaszQr) [inline modules/PriceBooks/actions/RelationAjax.php]: $queryGenerator->setFields(['purchase']);

File: modules/PriceBooks/models/RelationListView.php [hunk 4]
```diff
@@ -30,4 +30,20 @@ public function getHeaders()
...
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getLinks()
+	{
+		$relatedLink = parent::getLinks();
+		$relatedLink['RELATEDLIST_MASSACTIONS'][] = Vtiger_Link_Model::getInstanceFromValues([
+			'linktype' => 'RELATEDLIST_MASSACTIONS',
+			'linklabel' => 'LBL_MARGINP',
+			'linkurl' => 'javascript:Vtiger_RelatedList_Js.triggerMassMargin()',
```
Comment (TomaszQr) [inline modules/PriceBooks/models/RelationListView.php]: Vtiger_RelatedList_Js => 'Pricebooks_RelatedList_Js'

File: layouts/basic/modules/PriceBooks/Modals/SpecifyMargin.tpl [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+		<div class="col-12 row m-0">
+			<div class="fieldLabel col-sm-4 col-xl-4 pl-0 text-lg-left text-xl-right u-text-ellipsis">
+				<label class="text-right muted small font-weight-bold">
+					{\App\Language::translate('LBL_MARGINP')}
+				</label>
+			</div>
+			<div class="fieldValue col-sm-8 col-xl-8 px-0 px-sm-1">
+					<div class="input-group input-group-sm">
+							<input type="text" class="smallInputBox form-control form-control-sm js-margin"
+									data-validation-engine="validate[required,funcCall[Vtiger_NumberUserFormat_Validator_Js.invokeValidation]]" />
```
Comment (TomaszQr) [inline layouts/basic/modules/PriceBooks/Modals/SpecifyMargin.tpl]: data-validation-engine="validate[required,funcCall[Vtiger_NumberUserFormat_Validator_Js.invokeValidation]]" ? in js or in tpl?

File: modules/PriceBooks/actions/RelationAjax.php [hunk 6]
```diff
@@ -10,20 +10,55 @@
...
+	 * Setting margins.
+	 *
+	 * @param App\Request $request
+	 */
+	public function specifyMargin(App\Request $request)
 	{
-		$mode = $request->getMode();
-		if (!empty($mode) && method_exists($this, "$mode")) {
-			$this->$mode($request);
+		$margin = $request->getByType('margin', 'Double');
```
Comment (TomaszQr) [inline modules/PriceBooks/actions/RelationAjax.php]: Double =>  'NumberInUserFormat'

File: modules/PriceBooks/actions/RelationAjax.php [hunk 7]
```diff
@@ -10,20 +10,55 @@
...
+		$recordId = $request->getInteger('record');
+		$queryGenerator = static::getQuery($request);
+		$queryGenerator->setFields(['id', 'purchase']);
+		$currencyId = Vtiger_Record_Model::getInstanceById($recordId, 'PriceBooks')->get('currency_id');
+		$dbCommand = \App\Db::getInstance()->createCommand();
+		$dataReader = $queryGenerator->createQuery()->createCommand()->query();
+		while ($row = $dataReader->read()) {
+			if (!\App\Json::isEmpty($row['purchase'])) {
+				$purchase = \App\Json::decode($row['purchase']);
+				if (isset($purchase['currencies'][$currencyId]['price'])) {
```
Comment (rskrzypczak) [inline modules/PriceBooks/actions/RelationAjax.php]: ?

File: modules/PriceBooks/models/RelationListView.php [hunk 8]
```diff
@@ -30,4 +30,20 @@ public function getHeaders()
...
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getLinks()
+	{
+		$relatedLink = parent::getLinks();
+		$relatedLink['RELATEDLIST_MASSACTIONS'][] = Vtiger_Link_Model::getInstanceFromValues([
+			'linktype' => 'RELATEDLIST_MASSACTIONS',
+			'linklabel' => 'LBL_MARGINP',
```
Comment (rskrzypczak) [inline modules/PriceBooks/models/RelationListView.php]: ?

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 9]
```diff
@@ -1,34 +1,77 @@
...
+							} else {
+								element.validationEngine('hideAll');
+								AppConnector.request({
+									module: app.getModuleName(),
+									action: 'RelationAjax',
+									mode: 'specifyMargin',
+									relatedModule: self.moduleName,
+									selected_ids: selectedIds,
+									record: self.getParentId(),
+									margin: $('.js-margin').val()
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: margin: element.val()

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 10]
```diff
@@ -1,34 +1,77 @@
...
+Vtiger_RelatedList_Js(
+	"PriceBooks_RelatedList_Js",
+	{
+		triggerMassMargin: function() {
+			const self = Vtiger_RelatedList_Js.relatedListInstance;
+			if (self.checkListRecordSelected() !== true) {
+				let selectedIds = self.readSelectedIds(true);
+				app.showModalWindow({
+					url: 'index.php?module=PriceBooks&view=SpecifyMargin',
+					cb: () => {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: cb: (modalContainer) => {

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 11]
```diff
@@ -1,34 +1,77 @@
...
+						$('.js-modal__save').on('click', e => {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: modalContainer.find('.js-modal__save').on('click', e => {

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 12]
```diff
@@ -1,34 +1,77 @@
...
+							let element = $('.js-margin');
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: let element = modalContainer.find('.js-margin');

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 13]
```diff
@@ -1,34 +1,77 @@
...
-		this._super();
-		this.registerEditListPrice();
-	},
-});
+	{
+		registerEditListPrice: function() {
+			let thisInstance = this;
+			let element = this.content.find(".js-edit-listprice");
+			element.validationEngine(app.validationEngineOptions);
+			element.on("change", function(e) {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: element.on('change', function(e) {

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 14]
```diff
@@ -1,34 +1,77 @@
...
 	},
-	registerPostLoadEvents: function () {
-		this._super();
-		this.registerEditListPrice();
-	},
-});
+	{
+		registerEditListPrice: function() {
+			let thisInstance = this;
+			let element = this.content.find(".js-edit-listprice");
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: let element = this.content.find('.js-edit-listprice');

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 15]
```diff
@@ -1,34 +1,77 @@
...
+			element.validationEngine(app.validationEngineOptions);
+			element.on("change", function(e) {
+				e.stopPropagation();
+				let element = $(this);
+				element.formatNumber();
+				if (!element.validationEngine("validate")) {
+					AppConnector.request({
+						module: thisInstance.parentModuleName,
+						record: element.closest(".js-list__row").data("id"),
+						action: "RelationAjax",
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: ?

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 16]
```diff
@@ -1,34 +1,77 @@
...
+			let element = this.content.find(".js-edit-listprice");
+			element.validationEngine(app.validationEngineOptions);
+			element.on("change", function(e) {
+				e.stopPropagation();
+				let element = $(this);
+				element.formatNumber();
+				if (!element.validationEngine("validate")) {
+					AppConnector.request({
+						module: thisInstance.parentModuleName,
+						record: element.closest(".js-list__row").data("id"),
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: ?

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 17]
```diff
@@ -1,34 +1,77 @@
...
+						action: "RelationAjax",
+						mode: "addListPrice",
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: ?

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 18]
```diff
@@ -1,34 +1,77 @@
...
+						src_record: thisInstance.parentRecordId,
+						related_module: thisInstance.moduleName,
+						price: element.val()
+					}).done(function(responseData) {
+						if (responseData.result) {
+							Vtiger_Helper_Js.showPnotify({ text: app.vtranslate("JS_SAVE_NOTIFY_OK"), type: "success" });
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: ?

File: public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js [hunk 19]
```diff
@@ -1,34 +1,77 @@
...
-					action: 'RelationAjax',
-					mode: 'addListPrice',
-					src_record: thisInstance.parentRecordId,
-					related_module: thisInstance.moduleName,
-					price: element.val()
-				}).done(function (responseData) {
-					if (responseData.result) {
-						Vtiger_Helper_Js.showPnotify({text: app.vtranslate('JS_SAVE_NOTIFY_OK'), type: 'success'});
+Vtiger_RelatedList_Js(
+	"PriceBooks_RelatedList_Js",
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/PriceBooks/resources/RelatedList.js]: ?

**Pull request merged**